rust/arrow/src/array/union.rs

```diff
@@ -0,0 +1,1174 @@
+            .expect("Cannot be None")
+            .append(false)?;
+
+        self.type_id_builder.append(i8::default())?;
+
+        // Handle sparse union
+        if let None = self.value_offset_builder {
+            for (_, fd) in self.fields.iter_mut() {
+                fd.append_null_dynamic()?;
+            }
+        }
+        self.len += 1;
+        Ok(())
+    }
+
+    /// Appends a value to this builder.
+    pub fn append<T: ArrowPrimitiveType>(
+        &mut self,
+        type_name: &str,
+        v: T::Native,
+    ) -> Result<()> {
+        let type_name = type_name.to_string();
+
+        let mut field_data = match self.fields.remove(&type_name) {
+            Some(data) => data,
+            None => {
+                let field_data = match self.value_offset_builder {
+                    Some(_) => {
+                        FieldData::new(self.fields.len() as i8, T::get_data_type(), None)
+                    }
+                    None => {
+                        let mut fd = FieldData::new(
+                            self.fields.len() as i8,
+                            T::get_data_type(),
+                            Some(BooleanBufferBuilder::new(1)),
+                        );
+                        for _ in 0..self.len {
+                            fd.append_null::<T>()?;
+                        }
+                        fd
+                    }
+                };
+                field_data
+            }
+        };
+        self.type_id_builder.append(field_data.type_id)?;
+
+        match &mut self.value_offset_builder {
+            // Dense Union
+            Some(offset_builder) => {
+                offset_builder.append(field_data.slots as i32)?;
+            }
+            // Sparse Union
+            None => {
+                for (name, fd) in self.fields.iter_mut() {
+                    if name != &type_name {
+                        fd.append_null_dynamic()?;
+                    }
+                }
+            }
+        }
+        field_data.append_to_values_buffer::<T>(v)?;
+        self.fields.insert(type_name, field_data);
+
+        // Update the bitmap builder if it exists
+        if let Some(b) = &mut self.bitmap_builder {
+            b.append(true)?;
+        }
+        self.len += 1;
+        Ok(())
+    }
+
+    /// Builds this builder creating a new `UnionArray`.
+    pub fn build(mut self) -> Result<UnionArray> {
+        let type_id_buffer = self.type_id_builder.finish();
+        let value_offsets_buffer = self.value_offset_builder.map(|mut b| b.finish());
+        let mut children = Vec::new();
+        for (
+            name,
+            FieldData {
+                type_id,
+                data_type,
+                values_buffer,
+                slots,
+                bitmap_builder,
+                null_count,
+            },
+        ) in self.fields.into_iter()
+        {
+            let buffer = values_buffer.expect("UPDATE LATER").freeze();
```

Comment:
Reminder for the "UPDATE LATER", and TODO a few lines below

Reply:
Oh man.  Sorry about that, will fix.